mrp_subcontracting_purchase/models/stock_rule.py

```diff
@@ -0,0 +1,207 @@
+                    po_line_values.append(
+                        pol_obj._prepare_purchase_order_line_from_procurement(
+                            procurement.product_id,
+                            procurement.product_qty,
+                            procurement.product_uom,
+                            procurement.company_id,
+                            procurement.values,
+                            po,
+                        )
+                    )
+                    # Check if we need to advance the order date for the new line
+                    date_planned = procurement.values["date_planned"]
+                    order_date_planned = date_planned - relativedelta(
+                        days=procurement.values["supplier"].delay
+                    )
+                    if fields.Date.to_date(order_date_planned) < fields.Date.to_date(
+                        po.date_order
+                    ):
+                        po.date_order = order_date_planned
+            pol_obj.sudo().create(po_line_values)
+
+    @api.model
+    def _check_po_exists(self, domain, procurements, rules, company_id):
+        """Check if a PO exists for the current domain"""
+        po_obj = self.env["purchase.order"]
+        origins = {p.origin for p in procurements}
+        po = po_obj.sudo().search([dom for dom in domain], limit=1)
+        # Get the set of procurement origin for the current domain.
+        if not po:
+            positive_values = [
+                p.values
+                for p in procurements
+                if float_compare(
+                    p.product_qty, 0.0, precision_rounding=p.product_uom.rounding
+                )
+                >= 0
+            ]
+            if positive_values:
+                # We need a rule to generate the PO. However the rule generated
+                # the same domain for PO and the _prepare_purchase_order method
+                # should only uses the common rules's fields.
+                vals = rules[0]._prepare_purchase_order(
+                    company_id, origins, positive_values
+                )
+                # The company_id is the same for all procurements since
+                # _make_po_get_domain add the company in the domain.
+                # We use SUPERUSER_ID since
+                # we don't want the current user to be follower of the PO.
+                # Indeed, the current user may be a user without access to Purchase,
+                # or even be a portal user.
+                po = (
+                    po_obj.with_company(company_id).with_user(SUPERUSER_ID).create(vals)
+                )
+
+        else:
+            # If a purchase order is found, adapt its `origin` field.
+            if po.origin:
+                missing_origins = origins - set(po.origin.split(", "))
+                if missing_origins:
+                    po.write({"origin": po.origin + ", " + ", ".join(missing_origins)})
```

Comment:
Could you use `format` function for preparing string value?

Reply:
resolved